modules/uploads/scripts/prepare.ts

```diff
@@ -0,0 +1,167 @@
+import { RuntimeError, ScriptContext } from "../_gen/scripts/prepare.ts";
+import {
+	getMaxBytes,
+	PresignedUpload,
+	prismaToOutput,
+	UploadFile,
+} from "../utils/types.ts";
+import { getPresignedPutUrl, getPresignedMultipartUploadUrls } from "../utils/bucket.ts";
+import { DEFAULT_MAX_FILES_PER_UPLOAD } from "../config.ts";
+import { getS3EnvConfig } from "../utils/env.ts";
+import { DEFAULT_MULTIPART_CHUNK_SIZE } from "../config.ts";
+
+export interface Request {
+	userId?: string;
+	files: (UploadFile & { multipart: boolean })[];
+}
+
+export interface Response {
+	upload: PresignedUpload;
+}
+
+export async function run(
+	ctx: ScriptContext,
+	req: Request,
+): Promise<Response> {
+	const s3 = getS3EnvConfig();
+	if (!s3) throw new RuntimeError("s3_not_configured");
+
+	// Ensure there are files in the upload
+	if (req.files.length === 0) {
+		throw new RuntimeError("no_files");
+	}
+
+	// Ensure the number of files is within the limit
+	const maxFilesPerUpload = ctx.userConfig.maxFilesPerUpload ?? DEFAULT_MAX_FILES_PER_UPLOAD;
+	if (req.files.length > maxFilesPerUpload) {
+		throw new RuntimeError(
+			"too_many_files",
+			{
+				meta: {
+					reason: `Max files per upload is ${maxFilesPerUpload}` +
+					` (requested upload of ${req.files.length} files)`,
+				},
+			},
+		);
+	}
+
+	// Ensure there are no duplicate paths
+	const paths = new Set<string>();
+	const duplicates = new Set<string>();
+	for (const file of req.files) {
+		if (paths.has(file.path)) {
+			duplicates.add(file.path);
+		}
+		paths.add(file.path);
+	}
+	if (duplicates.size > 0) {
+		const duplicateString = Array.from(duplicates).map((path) =>
+			JSON.stringify(path)
+		).join(", ");
+		throw new RuntimeError("duplicate_paths", {
+			meta: {
+				reason: `Multiple files had paths ${duplicateString}`,
+			},
+		});
+	}
+
+	// Ensure the total content length is within the limit for single part uploads
+	const singlepartUploadContentLength = req.files.filter(f => !f.multipart).reduce(
+		(acc, file) => acc + BigInt(file.contentLength),
+		0n,
+	);
+	if (singlepartUploadContentLength > getMaxBytes(ctx.userConfig.maxUploadSize)) {
+		throw new RuntimeError(
+			"size_limit_exceeded",
+			{
+				meta: {
+					reason: `Max upload size is ${ctx.userConfig.maxUploadSize} bytes,` +
```

Comment:
The reason should be in the error description not the code. Just include the relevant metadata (ie maxUploadSize, singlePartUploadContentLength).